inference-engine/src/mkldnn_plugin/nodes/roll.cpp

```diff
@@ -0,0 +1,318 @@
+            }
+
+            if (axis < 0) {
+                axis += numberOfDims;
+            }
+
+            int64_t shift = shifts[index];
+            const int64_t currentDim = static_cast<int64_t>(m_shape[axis]);
+
+            if (shift >= currentDim || shift < -currentDim) {
+                IE_THROW() << "Shift: " << shift << " " << currentDim << " is out of bounds for Roll layer with name: " << layerName;
+            }
+
+            if (shift < 0) {
+                shift += currentDim;
+            }
+            axesToShift[static_cast<size_t>(axis)] += static_cast<size_t>(shift);
+        }
+        std::vector<AxisShift> result(axesToShift.size());
+        size_t resultIndex = 0;
+        // We have array sorted by axes
+        for (auto it = axesToShift.begin(); it != axesToShift.end(); ++it) {
+            result[resultIndex++] = {it->first, it->second};
+        }
+        return result;
+    }
+
+    template<typename T>
+    inline T *getElementByIndex(T *data, const std::vector<size_t> &counters, const std::vector<size_t> &strides) {
+        size_t offset = 0;
+        for (size_t idx = 0; idx < counters.size(); ++idx) {
+            offset += counters[idx] * strides[idx];
+        }
+        return data + offset;
+    }
+
+    template<typename T>
+    inline T *getElementByIndexWithShifts(T *data, const std::vector<size_t> &counters, const std::vector<size_t> &strides,
+                                          const std::vector<AxisShift> &axesShifts) {
+        size_t offset = 0;
+        auto it = axesShifts.begin();
+        for (size_t idx = 0; idx < counters.size(); ++idx) {
+            size_t current_axis_value = counters[idx];
+            if (it->axis == idx) {
+                current_axis_value += it->shift;
+                current_axis_value %= m_shape[idx];
+                ++it;
+            }
+            offset += current_axis_value * strides[idx];
+        }
+        return data + offset;
+    }
+
+    template<typename DataType>
+    void rollImpl(const Blob::CPtr &_input, const Blob::Ptr &_output) {
+        std::vector<AxisShift> axisShifts = calculateShifts(m_axes, m_shifts);
+
+        const auto *input =
+                _input->cbuffer().as<const DataType *>() + _input->getTensorDesc().getBlockingDesc().getOffsetPadding();
+        auto *output = _output->buffer().as<DataType *>() + _output->getTensorDesc().getBlockingDesc().getOffsetPadding();
+        const std::vector<size_t> strides = _input->getTensorDesc().getBlockingDesc().getStrides();
+        size_t workAmount = std::accumulate(m_shape.begin(), m_shape.end(), 1, std::multiplies<size_t>());
+        const size_t cache_line_per_tensor = std::max(1.0f, std::floor(static_cast<float>(workAmount) / mkldnn::utils::get_cache_size(3, false)));
+        int threads_number = std::min(parallel_get_max_threads(), static_cast<int>(cache_line_per_tensor));
+        parallel_nt(threads_number, [&](const int ithr, const int nthr) {
```

Comment:
Could you please compare performance of this implementation without parallelism versus ngraph reference implementation https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/reference/include/ngraph/runtime/reference/roll.hpp ?